chrome/thunderlink/content/thunderlink.js

```diff
@@ -44,43 +44,57 @@ var ThunderLinkChromeNS = {
     var appendToFileActive = prefService.getBoolPref("custom-tl-string-" + cstrnum + "-appendtofile-checkbox");
     console.log("ActivateCustomTlString: appendToFileActive: " + appendToFileActive + "\n");
 
+    var selectedMsgHeaders = gDBView.getSelectedMsgHdrs();
+    var selectionDelimiter = prefService.getCharPref("custom-tl-string-" + cstrnum + "-selection-delimiter");
+    var i = 0;
+
     // activate
     if (tagActive) {
-      this.TagEmail(prefService.getIntPref("custom-tl-string-" + cstrnum + "-tag"));
+      for (i = 0; i < selectedMsgHeaders.length; i++) {
+        this.TagEmail(prefService.getIntPref("custom-tl-string-" + cstrnum + "-tag"), selectedMsgHeaders[i]);
+      }
     }
+
     if (copyToClipboardActive) {
-      var procCustomTlStr = replaceVariables(customTlStr, gDBView.hdrForFirstSelectedMessage);
-      console.log("ActivateCustomTlString: procCustomTlStr resolved: " + procCustomTlStr + "\n");
+      var procCustomTlStr = "";
+      for (i = 0; i < selectedMsgHeaders.length; i++) {
+        procCustomTlStr += replaceVariables(customTlStr, selectedMsgHeaders[i]);
+        console.log("ActivateCustomTlString: procCustomTlStr resolved[" + i + "]: " + procCustomTlStr + "\n");
```

Comment:
This definitely works now - it generates the correct multi-header string, but the console.log message is no longer exactly right because with the header number in the message it means "I got this specific header" but the string is actually concatenating. You could just change the log message though - to something like `at header[" + i + "] - current string is " + procCustomTlStr = "...`